### ISSUE TITLE
Auto-load data and 'scene' (or state) on launch

### ISSUE THREAD

**Issue by llionevans** [Kitware/visualizer#47]
For me, a main motivation of using visualizer is to be able to share simulation results with a collaborator without them needing the hardware or software to open the data. Because of this we need to assume that the collaborator is also not familiar with ParaView.

What I'd like to do is preload a dataset in a saved state (e.g. using .pvsm). I've found documentation for visualizer regarding installation but not regarding usage. Could you please point me in the right direction?

Also, does visualizer accept all of the same input file types as ParaView? If not, is there a comprehensive list somewhere saying which types are compatible?

**Comment by jourdain** [comment]
Since Visualizer is using ParaView for the data processing and rendering, everything that ParaView can load, Visualizer can do it as well with only one caveat. Sometime for specific file extension, ParaView will ask you which reader you want to use to read a given file. Visualizer on the other hand will pick one by default without asking the user. If a specific one should be used, that is not the default, you can override it by providing a custom configuration like [here](https://github.com/Kitware/ParaView/blob/master/Web/Python/paraview/web/_default_proxies.py) using the [following argument](https://github.com/Kitware/visualizer/blob/master/server/pvw-visualizer.py#L44-L46).

To preload a file/state, you can use that [argument](https://github.com/Kitware/visualizer/blob/master/server/pvw-visualizer.py#L16)

Then you will use the launcher to control those arguments.

**Comment by jourdain** [comment]
It is definitely possible on the AMI as well. It is just that I'm not sure (don't remember) where the launcher config is stored as well as which version of PV+Visualizer that AMI has. You might be on the right track regarding the `launcher-TEMPLATE.json`.

But I want to make sure you solved all the pieces locally the way I'm thinking with the same version of Visualizer and ParaView.

- you edited your launcher-config.json to use some argument to load the data at startup.
- you "may need" to edit the index.html to process the URL arguments and add them to the configuration that get sent to the launcher. (It was automatic for some time, then that feature went away and I think it is back, but you may need to enable it manually)

**Comment by llionevans** [comment]
Locally I've only run the basic one session setup.
My hacky 'solution' is to have a shell script which I call and pass a variable to point at the .pvsm file

```
#!/bin/bash
./bin/pvpython   \
    ./share/paraview-5.4/web/visualizer/server/pvw-visualizer.py  \
    --content ./share/paraview-5.4/web/visualizer/www/               \
    --data $PWD/share/paraview-5.4/data                                      \
    --load-file $PATH_TO_DATA                                                      \
    --port 8080
```

I don't mind having a different index.html (index1.html, index2.html etc.) for each dataset if that's easier, because there'll only be a limited number of examples made available.

**Comment by llionevans** [comment]
@scottwittenburg 
I can't see a `pvw-launcher.conf` file, but there is a `/etc/init.d/pvw-start`.
There's nothing at all in `/etc/opt/`. On the ami, ParaView is under `/data/...

@jourdain
I'll look into using the launcher now.

**Comment by llionevans** [comment]
In the script here: [http://kitware.github.io/paraviewweb/docs/windows_10.html#Creating-the-ParaView-Web-Launcher-configuration](url)

Do I need to replace <paraview-root> and <apache-root> with the correct path, or are these variables?

**Comment by llionevans** [comment]
@scottwittenburg 
[Contents of pvw-start](https://pastebin.com/T7NbaKG0)

**Comment by jourdain** [comment]
In your use case (local), you don't need to worry about apache. But those <xxx> need to be replaced by you with something valid.

Moreover, since you are looking at the windows doc, make sure you fix the `sessionURL` with something like: `"sessionURL" : "ws://localhost:${port}/ws",`
That way you skip the Apache front-end.

**Comment by llionevans** [comment]
Ok here's my [launcher.config](https://pastebin.com/FkJ4RNjV), does that look ok?

**Comment by llionevans** [comment]
And modified [index.html](https://pastebin.com/ZbP2LwDu)

**Comment by jourdain** [comment]
Couple of comment

You can not use `~`, you need to use full path

```
"proxy_file" : "/tmp/proxy.txt",
```

Then remove line 13 otherwise you will connect twice...

**Comment by llionevans** [comment]
Ok I've updated to use a full path (and in index.html). By line 13 I assume you meant 13-16, i.e.
```
  "resources" : [ {
  "host" : "localhost",
  "port_range" : [9001, 9999]
  } ],
```

What should happen at this point? Do I need to do anything to launch before going to localhost via a browser? If I go to http://localhost:9000/?data=can.ex2 nothing happens. From launcher.config I assume that's the port I should go to.

**Comment by jourdain** [comment]
Yes, you need to start the launcher and yes the url that you used seems ok.

You can read more about the launcher [here](http://kitware.github.io/paraviewweb/docs/python_launcher.html) 

But the launcher is now provided by wslink (depending on the version of PV you are using), so the command line will be something along those lines:

```
./bin/pvpython lib/site-package/vtk/web/launcher.py launcher.config
```

or 

```
./bin/pvpython lib/site-package/wslink/launcher.py launcher.config
```

**Comment by llionevans** [comment]
Ok, I'm getting closer.

So the command I needed for v5.4 was:
`./bin/pvpython lib/site-package/python2.7/vtk/web/launcher.py launcher.config`

However, this threw back an error until I created a directory for `<paraview-root>/log`

I also had an error saying that 'resources' was missing from `launcher.config`, so I put back what I thought you wanted me to delete when you referred to line 13. I guess you actually meant `  },`. I hadn't realised that my first line was blank in what I put on pastebin previously. But deleting that returned an error too. So my `launcher.config` now looks like [this](https://pastebin.com/UALnPbKQ).

Now when I run the command above it doesn't return any text, the terminal just looks busy. Is this correct? When I go to http://localhost:9000/?data=can.ex2 on a browser I get a blank page. There's no error, the page title and favicon appear on top of the window but it's just blank.

**Comment by llionevans** [comment]
I've just checked the log file which says that it's not getting the data variable (see below).

2018-01-12 23:50:23,299:INFO:twisted:"127.0.0.1" - - [12/Jan/2018:23:50:23 +0000] "GET /?data=can.ex2 HTTP/1.1" 200 1518 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:57.0) Gecko/20100101 Firefox/57.0"
2018-01-12 23:50:23,535:INFO:twisted:"127.0.0.1" - - [12/Jan/2018:23:50:23 +0000] "GET /Visualizer.js HTTP/1.1" 304 - "http://localhost:9000/?data=can.ex2" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:57.0) Gecko/20100101 Firefox/57.0"
2018-01-12 23:50:24,461:ERROR:root:Some properties could not be resolved: ${dataDir}
2018-01-12 23:50:24,461:ERROR:root:Some properties could not be resolved: ${dataFile}
2018-01-12 23:50:24,486:ERROR:root:The command line failed
2018-01-12 23:50:24,486:ERROR:root:/home/ubuntu/ParaView/ParaView-5.4.0-Qt5-OpenGL2-MPI-Linux-64bit/bin/pvpython.exe -dr /home/ubuntu/ParaView/ParaView-5.4.0-Qt5-OpenGL2-MPI-Linux-64bit/share/paraview-5.4/web/visualizer/server/pvw-visualizer.py --port 9997 --authKey CHpdOTvCgOVoAqc9 --data ${dataDir} --load-file ${dataFile}
2018-01-12 23:50:24,487:INFO:twisted:"127.0.0.1" - - [12/Jan/2018:23:50:24 +0000] "POST /paraview/ HTTP/1.1" 503 384 "http://localhost:9000/?data=can.ex2" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:57.0) Gecko/20100101 Firefox/57.0"
2018-01-12 23:50:24,663:INFO:twisted:"127.0.0.1" - - [12/Jan/2018:23:50:24 +0000] "GET /ws HTTP/1.1" 404 145 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:57.0) Gecko/20100101 Firefox/57.0"`

**Comment by jourdain** [comment]
The line 13 was in index.html not the launcher... 
Otherwise you are asking your JavaScript to connect two times... 

You need to replace the section below 
```
<script>
      Visualizer.connect({ application: 'visualizer' });
      Visualizer.autoStopServer(10);
</script>
```

with that

```

<script>
// grab the parameter from the URL
function getUrlParameter(name) {
  name = name.replace(/[\[]/, '\\[').replace(/[\]]/, '\\]');
  var regex = new RegExp('[\\?&]' + name + '=([^&#]*)');
  var results = regex.exec(location.search);
  return results === null ? '' : decodeURIComponent(results[1].replace(/\+/g, ' '));
};
 
Visualizer.connect({
  application: 'visualizer',
  dataFile: getUrlParameter('data'),
  dataDir: '~/ParaView/ParaView-5.4.0-Qt5-OpenGL2-MPI-Linux-64bit/share/paraview-5.4/data'
});
</script>
```

You can not have both....

Moreover, your second `<script>` was never executed as it was outside the `<body></body>`.

**Comment by llionevans** [comment]
I realised that I hadn't removed the old part of index.html. I now get something more encouraging in the log file, but still not quite there.

018-01-13 00:02:47,102:INFO:twisted:"127.0.0.1" - - [13/Jan/2018:00:02:46 +0000] "GET /?data=can.ex2 HTTP/1.1" 200 1586 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:57.0) Gecko/20100101 Firefox/57.0"
2018-01-13 00:02:47,393:INFO:twisted:"127.0.0.1" - - [13/Jan/2018:00:02:46 +0000] "GET /Visualizer.js HTTP/1.1" 304 - "http://localhost:9000/?data=can.ex2" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:57.0) Gecko/20100101 Firefox/57.0"
2018-01-13 00:02:48,054:ERROR:root:The command line failed
2018-01-13 00:02:48,054:ERROR:root:/home/ubuntu/ParaView/ParaView-5.4.0-Qt5-OpenGL2-MPI-Linux-64bit/bin/pvpython.exe -dr /home/ubuntu/ParaView/ParaView-5.4.0-Qt5-OpenGL2-MPI-Linux-64bit/share/paraview-5.4/web/visualizer/server/pvw-visualizer.py --port 9996 --authKey ygnNE9VgEblf0URr --data /home/ubuntu/ParaView/ParaView-5.4.0-Qt5-OpenGL2-MPI-Linux-64bit/share/paraview-5.4/data --load-file can.ex2
2018-01-13 00:02:48,055:INFO:twisted:"127.0.0.1" - - [13/Jan/2018:00:02:47 +0000] "POST /paraview/ HTTP/1.1" 503 458 "http://localhost:9000/?data=can.ex2" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:57.0) Gecko/20100101 Firefox/57.0"
2018-01-13 00:02:48,083:INFO:twisted:"127.0.0.1" - - [13/Jan/2018:00:02:47 +0000] "GET /ws HTTP/1.1" 404 145 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:57.0) Gecko/20100101 Firefox/57.0"
2018-01-13 00:03:48,055:INFO:twisted:Timing out client: IPv4Address(TCP, '127.0.0.1', 39538)

**Comment by jourdain** [comment]
looks like the sessionURL may be wrong in your launcher config file.

That's the URL that the client use to connect to the pvpython process.

**Comment by llionevans** [comment]
For that I replaced
`    "sessionURL" : "ws://paraview:80/proxy?sessionId=${id}&path=ws",`
with
`    "sessionURL" : "ws://localhost:${port}/ws",`

**Comment by jourdain** [comment]
it should be fine, but that is the only thing that could explain why the client does not connect.

Unless your load time of paraview due to the file you are loading is bigger than `"timeout" : 30,` => 30 seconds.

**Comment by llionevans** [comment]
I've increased the timeout but it doesn't seem to make a difference. But if you look at the log, the error occurs only one second after the page is called.

2018-01-13 00:34:21,020:INFO:twisted:"127.0.0.1" - - [13/Jan/2018:00:34:20 +0000] "GET /?data=can.ex2
2018-01-13 00:34:22,201:ERROR:root:The command line failed

I'll do a copy of my latest config and index files just in case I've done something stupid.

**Comment by scottwittenburg** [comment]
Do you see process-specific log files in your log directory?  If so, perhaps those will have some information about why the command line is failing.

**Comment by llionevans** [comment]
[launcher.config](https://pastebin.com/uJcYXJ1A) & [index.html](https://pastebin.com/nDrpze5z)

**Comment by llionevans** [comment]
@scottwittenburg 
There are lots of logs named similar to `7fd894ba-f7f9-11e7-8a5f-08002705f961.txt` but they're all empty.

**Comment by scottwittenburg** [comment]
You might try copying the command line that is being run out of your log file and pasting it into a terminal to see what happens when you run it by hand.

**Comment by llionevans** [comment]
Hah, genius! It didn't like the .exe in the call because we're no longer in windows. It now works once I removed `.exe`.

Thanks so much for your patience and support. I better get to my weekend with the family before I get in trouble. 

Now that I understand what's happening much more I'll have a play with implementing what I had in mind on the ami first thing on Monday.

**Comment by scottwittenburg** [comment]
Glad you got it working, have a nice weekend 😄

**Comment by llionevans** [comment]
And you two too.
~5 hrs before little footsteps come to wake me up.

**Comment by llionevans** [comment]
Hi both, hope you had a good weekend. I had a bit of a chance to look at getting data to pre-load from the url using EC2 today.

The scripts for v4.4 on the ami look quite different to what I was looking at last week for v5.4.

I assumed that `index.html` was the best starting point. The first thing I tried to do was hard code it in to get rid of the chance that I wasn't doing `getUrlParameter` correctly.

Near the bottom I found the following lines:
`<script type="text/javascript">`
`    var config = {`
`        sessionManagerURL:vtkWeb.properties.sessionManagerURL,`
`        application: "visualizer"`
`    },`

So I tried adding:
`        dataFile:"can.ex2"`

I also tried a few variations e.g. without speechmarks, using the variable `fileselector` which I found in `main.js`. But none seemed to do anything. I restarted the web server between each change, so I could see that the source code was successfully changing.

Any suggestions would be gratefully received.

**Comment by jourdain** [comment]
Well, if you don't change anything in the launcher config, I'm not expecting anything to preload.
As no additional argument will be passed to the visualizer python script...

**Comment by llionevans** [comment]
Ok, so I tried adding `"--load-file", "${dataFile}"` to line 30 of `launcher.json`. What else would I need to do?

**Comment by jourdain** [comment]
That's the trick. Did you change it in the config that actually run (the template one + restart?)

**Comment by llionevans** [comment]
No, I hadn't. I did it in the wrong one! Just tried again in template and hard coding can.ex2 into index.html worked. Thanks.

Now to try adding getUrlParameter.

**Comment by llionevans** [comment]
Perfect, it worked. Hopefully by the end of the week I'll have added a few small demo datasets. Then If you're interested I can send you the link to have a look.

Thanks again for the help.

**Comment by llionevans** [comment]
I'm thinking about uploading my datasets to a separate (always up) server, then mounting this on the EC2 instance running visualizer.

My assumption is that this is a bad idea because ParaView expects datasets to potentially be very large and therefore doesn't do unscalable things like make temporary copies or read the whole thing into RAM. If my assumption is correct I'll be depending on how fast the network can download each timestep file, which probably means it would be better to store all data on my EC2 instance to avoid a major hit to performance?

I just wanted to ask for your opinion because I wasn't sure how ParaView handles opened data.

**Comment by jourdain** [comment]
You should put your dataset on an EBS which can then be only mounted on a system at a time.
But that way, you will have better load time as ParaView expect to have those file locally available.
But you will only have one server available with the data at a given time.

If you want both, you can put your data on a server that is always up. But then you have a cron job on the ec2-gpu machine that rsynch them back to the SSD drive of that machine periodically. That way those become local to the ParaView and the read time will always be good.

**Comment by llionevans** [comment]
Using rsync on boot of the viz server is a great idea, thanks.

**Comment by llionevans** [comment]
Yes, my data will likely be pretty large. Is vtk.js better adapted for larger data? Would I need to convert data to the vtk format?

Other than having a few demo datasets as an example of research I think the main way I was planning to use visualizer was to share research data with collaborators in such a way that they don't need to install software (and may not have appropriate hardware).

**Comment by llionevans** [comment]
It's great because I can visualize both CT and FEA data using the same program.

Yes, I think the issue is present on v5.4 because I tried on my local machine before uploading to AWS and that's where I first noticed it. I'll try looking in a bit more detail to make sure, if I'm correct I'll start a new issue.
